### PR TITLE
Add PFL support

### LINKED ISSUE #79: Consider not tracklisting if a track is faded down
Problem: then we'd have to check if it hasn't been tracklisted each time it gets faded up - not too much of a faff, but it is a bit spaghetti

### PR DESCRIPTION
**Pull request by mstratford**
Finally! PFL support for WebStudio!

- Turn it on and off per channel at any time.
- PFL'ing when not playing will play with fader off, as would the Sonifex.
- Tracklisting now is now fader aware, it will only tracklist at the start of a song having it's fader turned on. Fixes #79

Todo:
- [x] Add a clear PFL button.
- [x] Hide PFL button if custom outputs applied (it won't work)
- [x] Maybe headphone volume monitor meter / volume control?

### CODE REVIEW

File: src/mixer/audio.ts [hunk 1]
```diff
@@ -156,15 +166,30 @@ class Player extends ((PlayerEmitter as unknown) as { new (): EventEmitter }) {
...
     } else {
-      this.wavesurfer.setVolume(1);
-      if (!this.customOutput) {
-        (this.wavesurfer as any).backend.gainNode.gain.value = linear;
+      try {
+        (this.wavesurfer as any).backend.analyser.disconnect(
+          this.engine.headphonesNode
+        );
+      } catch (e) {
+        console.log("Didn't disconnect.");
```
Comment (markspolakovs) [inline src/mixer/audio.ts:192]: Not a fan of swallowing exceptions, this should at least get sent to Raygun (which a `console.warn(e)` would do).

File: src/mixer/audio.ts [hunk 2]
```diff
@@ -327,6 +354,9 @@ export class AudioEngine extends ((EngineEmitter as unknown) as {
   newsEndCountdownEl: HTMLAudioElement;
   newsEndCountdownNode: MediaElementAudioSourceNode;
 
+  // Headphones
+  headphonesNode: GainNode;
```
Comment (markspolakovs) [inline src/mixer/audio.ts:358]: Possibly a more generic name? e.g. for multi-person shows (lol) I'd be sending guests the "headphones feed"

File: src/mixer/audio.ts [hunk 3]
```diff
@@ -415,22 +448,40 @@ export class AudioEngine extends ((EngineEmitter as unknown) as {
...
     this.newsStartCountdownNode.connect(this.audioContext.destination);
     this.newsEndCountdownNode.connect(this.audioContext.destination);
+
+    // Send the headphones feed to the headphones.
+    const db = -12; // DB gain on headphones (-6 to match default trim)
+    this.headphonesNode.gain.value = Math.pow(10, db / 20);
+    this.headphonesNode.connect(this.audioContext.destination);
+  }
+
+  _connectFinalCompressor(headphones: boolean) {
```
Comment (markspolakovs) [inline src/mixer/audio.ts:466]: `headphones` isn't really clear on what it does - please either rename it or add a doc comment explaining its semantics

File: src/mixer/audio.ts [hunk 4]
```diff
@@ -452,6 +503,30 @@ export class AudioEngine extends ((EngineEmitter as unknown) as {
...
+  public setPFL(number: number, enabled: boolean) {
+    var routeMainOut = true;
+    var player = this.getPlayer(number);
+
+    if (player) {
+      player.setPFL(enabled);
+    }
+
+    var i;
+    for (i = 0; i < this.players.length; i++) {
```
Comment (markspolakovs) [inline src/mixer/audio.ts:515]:
```suggestion
    for (let i = 0; i < this.players.length; i++) {
```

File: src/mixer/audio.ts [hunk 5]
```diff
@@ -452,6 +503,30 @@ export class AudioEngine extends ((EngineEmitter as unknown) as {
...
+        // PFL is enabled on this channel, so we're not routing the regular output to H/Ps.
+        routeMainOut = false;
+        console.log("Player", i, "is PFL'd.");
+      } else {
+        console.log("Player", i, "isn't PFL'd.");
+      }
+    }
+    console.log("Routing main out?", routeMainOut);
+
+    this._connectFinalCompressor(routeMainOut);
```
Comment (markspolakovs) [inline src/mixer/audio.ts:527]: Er, I don't quite fully understand how this works. The way I read it, when a single player is PFL'd, we don't route the final compressor to headphones - which is fair, but where do the other players get routed then (assuming only one is PFL'd)?

**Comment by mstratford** [comment]
Leaving out Split PFL for now, it's more complicated.

I'm seeing some weird graphical glitches with the VUs after using PFL.

File: src/mixer/audio.ts (hunk 3 above)
Comment (mstratford) [reply src/mixer/audio.ts]: Did both 👍

File: src/mixer/audio.ts (hunk 1 above)
Comment (mstratford) [reply src/mixer/audio.ts]: Relabled to explain why.

File: src/mixer/audio.ts (hunk 2 above)
Comment (mstratford) [reply src/mixer/audio.ts:357]: Hmm, I was expecting anything going to the headphones to go through this node, so I thought this _was_ generic...

**Comment by mstratford** [comment]
PFL levels need to be calibrated, but I want to do this separately with input from webstudio-dev testers first :)

File: src/mixer/audio.ts [hunk 6]
```diff
@@ -156,15 +166,28 @@ class Player extends ((PlayerEmitter as unknown) as { new (): EventEmitter }) {
...
     } else {
-      this.wavesurfer.setVolume(1);
-      if (!this.customOutput) {
-        (this.wavesurfer as any).backend.gainNode.gain.value = linear;
+      try {
+        (this.wavesurfer as any).backend.analyser.disconnect(
+          this.engine.headphonesNode
+        );
+      } catch (e) {
+        // This connection wasn't connected anyway, ignore.
```
Comment (markspolakovs) [inline src/mixer/audio.ts:190]: Worth at least a `console.warn()`?

File: src/mixer/audio.ts (hunk 2 above)
Comment (markspolakovs) [reply src/mixer/audio.ts:357]: Fair.

**Review by markspolakovs — APPROVED** [review]
Seems largely sane

File: src/mixer/audio.ts (hunk 6 above)
Comment (mstratford) [reply src/mixer/audio.ts:190]: Hmm. I could do, but it is kinda expected behaviour due to the way it's written.

**Pull request merged**